Type inference should be more careful when unifying wildcards to avoid unsoundness

```scala
class Foo[T](var elem: T)

object Test {
  def setFirstInPair[T](pair: (Foo[T], Foo[T])) = {
    pair._1.elem = pair._2.elem
  }

  def setFirstInList[T](list: List[Foo[T]]) = {
    list(0).elem = list(1).elem
  }

  def test1(): Unit = {
    val fooInt = new Foo(1)
    val fooString = new Foo("")
    val pair: (Foo[_], Foo[_]) = (fooInt, fooString)
    setFirstInPair(pair) // Should be an error, like in Scala 2
    println(fooInt.elem + 1) // ClassCastException in Dotty
  }

  def test2(): Unit = {
    val fooInt = new Foo(1)
    val fooString = new Foo("")
    val list: List[Foo[_]] = List(fooInt, fooString)
    setFirstInList(list) // Should be an error, like in Scala 2
    println(fooInt.elem + 1) // ClassCastException in Dotty
  }

  def main(args: Array[String]): Unit = {
    test1()
    test2()
  }
}
```
The Scala 2 errors are:
```scala
try/sound.scala:16: error: no type parameters for method setFirstInPair: (pair: (Foo[T], Foo[T]))Unit exist so that it can be applied to arguments ((Foo[_], Foo[_]))
 --- because ---
argument expression's type is not compatible with formal parameter type;
 found   : (Foo[_], Foo[_])
 required: (Foo[?T], Foo[?T])
    setFirstInPair(pair) // Should be an error, like in Scala 2
    ^
try/sound.scala:16: error: type mismatch;
 found   : (Foo[_], Foo[_])
 required: (Foo[T], Foo[T])
    setFirstInPair(pair) // Should be an error, like in Scala 2
                   ^
try/sound.scala:24: error: no type parameters for method setFirstInList: (list: List[Foo[T]])Unit exist so that it can be applied to arguments (List[Foo[_]])
 --- because ---
argument expression's type is not compatible with formal parameter type;
 found   : List[Foo[_]]
 required: List[Foo[?T]]
    setFirstInList(list) // Should be an error, like in Scala 2
    ^
try/sound.scala:24: error: type mismatch;
 found   : List[Foo[_]]
 required: List[Foo[T]]
    setFirstInList(list) // Should be an error, like in Scala 2
                   ^
```

Note that the following compiles fine in Scala 2, so it's not just applying a blanket ban on unifying with wildcards:
```scala
  def foo[T](foo: Foo[T]) = {}
  val x: Foo[_] = new Foo(1)
  foo(x)
```

I'm actually not sure how to reason about the difference in Dotty since we don't have a concept of existential types, /cc @Blaisorblade ?

Turns out we don't even need wildcards to have a problem:
```scala
abstract class Foo {
  type A
  var elem: A
}

object Test {
  def setFirstInList[T](list: List[Foo { type A = T }]) = {
    list(0).elem = list(1).elem
  }

  def main(args: Array[String]): Unit = {
    val fooInt = new Foo { type A = Int; var elem = 1 }
    val fooString = new Foo { type A = String; var elem = "" }
    val list: List[Foo] = List(fooInt, fooString)
    setFirstInList(list) // Should be an error, like in Scala 2
    println(fooInt.elem + 1) // ClassCastException in Dotty
  }
}
```
Scala 2 says:
```scala
try/sound2.scala:15: error: type mismatch;
 found   : List[Foo]
 required: List[Foo{type A = this.A}]
    setFirstInList(list) // Should be an error, like in Scala 2
                   ^
```

Whereas Dotty happily infers:
```scala
setFirstInList[Foo#A](list)
```

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -1148,7 +1148,7 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
       def matchAbstractTypeMember(info1: Type) = info1 match {
         case TypeBounds(lo, hi) if lo ne hi =>
           tp2.refinedInfo match {
-            case rinfo2: TypeBounds if tp1.widenExpr.isSingleton =>
+            case rinfo2: TypeBounds if tp1.isStable =>
               val ref1 = tp1.widenExpr.select(name)
               isSubType(rinfo2.lo, ref1) && isSubType(ref1, rinfo2.hi)
             case _ =>

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -1132,7 +1132,7 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
     /*>|>*/ trace(i"hasMatchingMember($tp1 . $name :? ${tp2.refinedInfo}), mbr: ${tp1.member(name).info}", subtyping) /*<|<*/ {
       val rinfo2 = tp2.refinedInfo
 
-      // If the member is an abstract type, compare the member itself
+      // If the member is an abstract type and the prefix is a path, compare the member itself
       // instead of its bounds. This case is needed situations like:
       //
       //    class C { type T }
@@ -1148,7 +1148,7 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
       def matchAbstractTypeMember(info1: Type) = info1 match {
         case TypeBounds(lo, hi) if lo ne hi =>
           tp2.refinedInfo match {
-            case rinfo2: TypeBounds =>
+            case rinfo2: TypeBounds if tp1.widenExpr.isSingleton =>
               val ref1 = tp1.widenExpr.select(name)
               isSubType(rinfo2.lo, ref1) && isSubType(ref1, rinfo2.hi)
             case _ =>

--- a/tests/neg/i5948.scala
+++ b/tests/neg/i5948.scala
@@ -1,0 +1,20 @@
+abstract class Foo {
+  type A
+  var elem: A
+}
+
+object Test {
+  def setFirstInList[T](list: List[Foo { type A = T }]) = {
+    list(0).elem = list(1).elem
+  }
+
+  def main(args: Array[String]): Unit = {
+    val fooInt = new Foo { type A = Int; var elem = 1 }
+    val fooString = new Foo { type A = String; var elem = "" }
+
+    val list: List[Foo] = List(fooInt, fooString)
+    setFirstInList[Foo#A](list) // error
+    setFirstInList(list) // error
+    println(fooInt.elem + 1)
+  }
+}

--- a/tests/neg/i5948a.scala
+++ b/tests/neg/i5948a.scala
@@ -1,0 +1,14 @@
+trait Bar {
+  type A
+}
+
+object Test {
+  def test0: Unit = {
+    val b1: Bar = new Bar {}
+    val b2: Bar { type A = Bar#A } = b1 // error
+  }
+  def test1: Unit = {
+    val b1: List[Bar] = List(new Bar {})
+    val b2: List[Bar { type A = Bar#A }] = b1 // error
+  }
+}

